### ISSUE TITLE
Question about codebooks

### ISSUE THREAD

**Issue by dyhan316** [935963004/LaBraM#47]
Thank you so much for the wonderful paper! 

We're trying to do a similar thing with our ECoG data, and had a few questions if you don't mind regarding the tokenizer. 

1. **Why did you only tokenize the output?** : In the appendix, you show that making the model predict the codes is better than predicting the raw amplitude and phase). Since the tokenizer seems to be beneficial, we were wondering if you tried using the code themselves as inputs? Why or why not? 

2. **Is the output of the decoder one of 8192? ** : I'm not sure if I understand correctly, but since the codebook size is 8192, the variety out reconstructions that we can obtain from the decoders 8192? Is my assessment correct? 

3. **Do you think that a separate tokenizer must be trained for each modality? (ex : EEG/iEEG/MEG) ** Also, why did you use the transformer for the tokenizer? Wouldn't that kind of destroy the temporal information compared to models like CNN? 

4. In Figure 7, you showed the reconstructed amplitude and phase. We were wondering: woul d it be possible the use those reconstructed amplitude and phase to get the reconstruction of the original EEG signal itself? We were wondering if you could've used the recontstructed EEG signal as the loss instead of the phase and amplitude? (i.e. the model predicts the tokens ==> the tokens are changed to the amplitude and phase ==> this is further converted to the reconstruction of the EEG single ==> take MSE with it)

Thank you in advance for your response :)

**Comment by 935963004** [comment]
1. This is an interesting question. The tokenizer can be regarded as a denoiser as EEG signals are noisy. We input noisy EEG signals, and then get clean codes. Those codes, although clean, may lose information such as correlations between different patches, as they are independent. Therefore, we use the codes to serve as the target when performing masked EEG modeling like BERT. We have not tried use the codes as input directly, but in my view, this way may not get satisfactory results.
2. You are correct.
3. I think so. If you want to train it with multiple modalities, maybe disentangling learning would be beneficial. It is okay to only use CNN, but in this work, we want the architecture to be consistent and it does work using transformer.
4. It is possible, but reconstructing the amplitude and phase can be easier for the model the learn when your data is insufficient and your model is somewhat small. We did this ablation in Appendix.

**Comment by dyhan316** [comment]
Thank you so much for your response. I just have one more further question regarding 2. if you don't mind. 

2. To be precise the number of latents available that the decoder uses from (for each input token) is 8192, but the output of the decoder itself has much more possibilities right? (since the decoder mixes the codebooks from all channels for four tokens (4s x number of channels)

**Comment by 935963004** [comment]
Right. The tokenizer is greatly inspired by VQ-VAE. If you want to know more about it, you are encouraged to refer to the original VQ-VAE paper.
